[dbt tutorial] downstream assets section of tutorial (#15597)
dagster-io/dagster: docs/content/integrations/dbt/using-dbt-with-dagster/downstream-assets.mdx

  tutorial.
</Note>

By this point, you've [set up a dbt project](/integrations/dbt/using-dbt-with-dagster/set-up-dbt-project), [loaded dbt models into Dagster as assets](/integrations/dbt/using-dbt-with-dagster/load-dbt-models), and [created assets upstream of your dbt models](/integrations/dbt/using-dbt-with-dagster/upstream-assets).

In this step, you'll:

- [Install the plotly library](#step-1-create-the-order_count_chart-asset)
- [Create a downstream asset that computes a chart using plotly](#step-2-create-the-order_count_chart-asset)
- [Materialize the `order_count_chart` asset](#step-3-materialize-the-order_count_chart-asset)

---

## Step 1: Install the plotly library

```shell
pip install plotly
```

## Step 2: Create the order_count_chart asset

> You've added upstream assets to your data pipeline, but nothing downstream - until now. In this step, you'll create a Dagster asset called `order_count_chart` that uses the data in the `customers` dbt model to computes a plotly chart of the number of orders per customer.

Reviewer: data pipeline needs to be clarified. 

```suggestion
You've added upstream assets to your dbt assets, but nothing downstream - until now. In this step, you'll create a Dagster asset called `order_count_chart` that uses the data in the `customers` dbt model to computes a plotly chart of the number of orders per customer.
```

   import pandas as pd
   import plotly.express as px
   from dagster_dbt import (
       DbtCliResource,
       build_schedule_from_dbt_selection,
       dbt_assets,
       get_asset_key_for_model,
   )

   from dagster import Definitions, MetadataValue, OpExecutionContext, asset
   ```

   This adds an import for plotly, as well as <PyObject module="dagster_dbt" object="get_asset_key_for_model" /> and <PyObject module="dagster" object="MetadataValue" />, which we'll use in our asset.

2. Before `defs = Definitions(`, add the definition for the `order_count_chart` asset:

   ```python file=integrations/dbt/tutorial/downstream_assets/definitions.py startafter=start_downstream_asset endbefore=end_downstream_asset
   @asset(deps=get_asset_key_for_model([jaffle_shop_dbt_assets], "customers"))
   def order_count_chart(context):
       # read the contents of the customers table into a Pandas DataFrame
       connection = duckdb.connect(os.fspath(duckdb_database_path))
       customers = connection.sql("select * from customers").df()

       # create a plot of number of orders by customer and write it out to an HTML file
       fig = px.histogram(customers, x="number_of_orders")

>        fig.update_layout(bargap=0.2)

Reviewer: Is `update_layout` really necessary?

Author: very possibly not, but this was the code that was here before, so my bias is to leave it for now to minimize the number of moving pieces

       save_chart_path = duckdb_database_path.parent.joinpath("order_count_chart.html")
       fig.write_html(save_chart_path, auto_open=True)

       # tell Dagster about the location of the HTML file, so it's easy to access from the Dagster UI
       context.add_output_metadata(
           {"plot_url": MetadataValue.url("file://" + os.fspath(save_chart_path))}
       )
   ```

   This asset definition looks similar the asset we defined in the previous section. In this case, instead of fetching data from an external source and writing it to DuckDB, it reads data from DuckDB, and then uses it to make a plot.

>    The line `deps=get_asset_key_for_model([jaffle_shop_dbt_assets], "customers")` tells Dagster that this asset is downstream of the `customers` dbt model. The consequence of this is that it will be displayed as such in Dagster's UI, and, if you launch a run to materialize both of them, Dagster won't run `order_count_chart` until `customers` completes.

Reviewer: Consequence is ambiguous, and I cannot tell if this is a bad thing or good thing. Let's just use a different word.

```suggestion
   The line `deps=get_asset_key_for_model([jaffle_shop_dbt_assets], "customers")` tells Dagster that this asset is downstream of the `customers` dbt model. This dependency will be displayed in Dagster's UI. If you launch a run that materializes both assets, Dagster won't run `order_count_chart` until `customers` completes.
```

Author: 👍
